ambari-server/src/main/java/org/apache/ambari/server/serveraction/kerberos/KerberosServerAction.java

```diff
@@ -449,13 +448,20 @@ protected CommandReport processIdentities(Map<String, Object> requestSharedDataC
       }
 
       try {
-        final Map<String, ? extends Collection<String>> serviceComponentFilter = (pruneServiceFilter())
+        final Map<String, Collection<String>> serviceComponentFilter = (pruneServiceFilter())
             ? kerberosKeytabController.adjustServiceComponentFilter(clusters.getCluster(getClusterName()), true, getServiceComponentFilter())
             : getServiceComponentFilter();
-        final Collection<KerberosIdentityDescriptor> serviceIdentities = serviceComponentFilter == null ? null : calculateServiceIdentities(getClusterName(), serviceComponentFilter);
-        for (ResolvedKerberosKeytab rkk : kerberosKeytabController.getFilteredKeytabs(serviceComponentFilter, getHostFilter(), getIdentityFilter())) {
+        final Collection<KerberosIdentityDescriptor> serviceIdentities = serviceComponentFilter == null ? null : kerberosKeytabController.calculateServiceIdentities(getClusterName(), serviceComponentFilter);
+        final Set<ResolvedKerberosKeytab> filteredKeytabs = kerberosKeytabController.getFilteredKeytabs(serviceComponentFilter, getHostFilter(), getIdentityFilter());
```

Comment:
It is not clear why this is not pulling back all of the relevant `ResolvedKerberosKeytab` items.  If the Kerberos descriptor for TEZ_CLIENT indicates that it needs the HDFS headless keytab file, there should be a set of records in the Ambari database mapping the HDFS principal name and keytab file to TEZ_CLIENT.   If so, then this filter should return the correct items. If not, maybe the problem is there.

Reply:
```
SELECT * FROM kerberos_keytab_principal inner join kkp_mapping_service using (kkp_id) where principal_name='hdfs-c1@EXAMPLE.COM';
```
```
59	/etc/security/keytabs/hdfs.headless.keytab	hdfs-c1@EXAMPLE.COM	1	1	HDFS	NAMENODE
59	/etc/security/keytabs/hdfs.headless.keytab	hdfs-c1@EXAMPLE.COM	1	1	HDFS	HDFS_CLIENT
59	/etc/security/keytabs/hdfs.headless.keytab	hdfs-c1@EXAMPLE.COM	1	1	MAPREDUCE2	HISTORYSERVER
59	/etc/security/keytabs/hdfs.headless.keytab	hdfs-c1@EXAMPLE.COM	1	1	TEZ	TEZ_CLIENT
59	/etc/security/keytabs/hdfs.headless.keytab	hdfs-c1@EXAMPLE.COM	1	1	YARN	APP_TIMELINE_SERVER
59	/etc/security/keytabs/hdfs.headless.keytab	hdfs-c1@EXAMPLE.COM	1	1	YARN	TIMELINE_READER
```

Reply:
Because it simply queries the DB (where we have no information about the Kerberos descriptor unless the user submitted one) based on the service filter we passed (i.e HDFS):
```
SELECT kkp.*, h.host_name, kkpm.service_name, kkpm.component_name
FROM kerberos_keytab_principal kkp, hosts h, kkp_mapping_service kkpm
WHERE kkp.host_id = h.host_id 
AND kkp.kkp_id = kkpm.kkp_id
AND kkpm.service_name = 'HDFS' 
AND h.host_name = 'c7403.ambari.apache.org'
ORDER BY h.host_name, kkpm.service_name, kkpm.component_name;
```
Returns:

kkp_id|keytab_path|principal_name|host_id|is_distributed|host_name|service_name|component_name
-|-|-|-|-|-|-|-
6|/etc/security/keytabs/spnego.service.keytab|HTTP/c7403.ambari.apache.org@AMBARI.APACHE.ORG|3|1|c7403.ambari.apache.org|HDFS|DATANODE
30|/etc/security/keytabs/smokeuser.headless.keytab|ambari-qa-cluster1@AMBARI.APACHE.ORG|3|1|c7403.ambari.apache.org|HDFS|DATANODE
37|/etc/security/keytabs/dn.service.keytab|dn/c7403.ambari.apache.org@AMBARI.APACHE.ORG|3|1|c7403.ambari.apache.org|HDFS|DATANODE

This is why it would only regenerate DN's keytab.

However we know that other service components needs the headless keytab:
```
SELECT kkp.*, h.host_name, kkpm.service_name, kkpm.component_name
FROM kerberos_keytab_principal kkp, hosts h, kkp_mapping_service kkpm
WHERE kkp.host_id = h.host_id 
AND kkp.kkp_id = kkpm.kkp_id
AND kkp.principal_name = 'hdfs-cluster1@AMBARI.APACHE.ORG' 
AND h.host_name = 'c7403.ambari.apache.org'
ORDER BY h.host_name, kkpm.service_name, kkpm.component_name
```
Returns:

kkp_id|keytab_path|principal_name|host_id|is_distributed|host_name|service_name|component_name
-|-|-|-|-|-|-|-
15|/etc/security/keytabs/hdfs.headless.keytab|hdfs-cluster1@AMBARI.APACHE.ORG|3|1|c7403.ambari.apache.org|TEZ|TEZ_CLIENT
15|/etc/security/keytabs/hdfs.headless.keytab|hdfs-cluster1@AMBARI.APACHE.ORG|3|1|c7403.ambari.apache.org|AMBARI_METRICS|METRICS_COLLECTOR

This is why we need to add other services; let me think it over again; there may be a more elegant and easy way to solve this issue.